src/packages/client/src/runtime/getPrismaClient.ts

```diff
@@ -907,15 +935,107 @@ new PrismaClient({
       )
     }
 
-    async $transaction(promises: Array<any>): Promise<any> {
+    /**
+     * @deprecated
+     */
+    async $___transaction(promises: Array<any>): Promise<any> {
       try {
-        return this.$transactionInternal(promises)
+        return this.$___transactionInternal(promises)
       } catch (e) {
         e.clientVersion = this._clientVersion
         throw e
       }
     }
 
+    /**
+     * Execute queries within a transaction
+     * @param input a callback or a query list
+     * @param options to set timeouts
+     * @returns
+     */
+    $transaction(input: any, options?: any) {
+      // eslint-disable-next-line prettier/prettier
+      if (!this._hasPreviewFlag('longRunningTransactions')) {
+        return this.$___transaction(input)
+      }
+
+      try {
+        return this._transaction(input, options)
+      } catch (e) {
+        e.clientVersion = this._clientVersion
+
+        throw e
+      }
+    }
+
+    /**
+     * Decide upon which transaction logic to use
+     * @param input
+     * @param options
+     * @returns
+     */
+    private async _transaction(input: any, options?: any) {
+      if (typeof input === 'function') {
+        return this._transactionWithCallback(input, options)
+      }
+
+      return this._transactionWithRequests(input, options)
+    }
+
+    /**
+     * Perform a long-running transaction
+     * @param callback
+     * @param options
+     * @returns
+     */
+    private async _transactionWithCallback(
+      callback: (client: Client) => Promise<unknown>,
+      options?: { maxWait: number; timeout: number },
+    ) {
+      // we ask the query engine to open a transaction
+      const info = await this._engine.transaction('start', options)
+
+      let result: unknown
+      try {
+        // execute user logic with a proxied the client
+        result = await callback(transactionProxy(this, info.id))
+
+        // it went well, then we commit the transaction
+        await this._engine.transaction('commit', info)
+      } catch (e) {
+        // it went bad, then we rollback the transaction
+        await this._engine.transaction('rollback', info)
+
+        throw e
+      }
+
+      return result
+    }
+
+    /**
+     * Execute a batch of requests in a transaction
+     * @param requests
+     * @param options
+     */
+    private async _transactionWithRequests(
+      requests: Array<Promise<unknown>>,
+      options?: { maxWait: number; timeout: number },
+    ) {
+      return this._transactionWithCallback(async (prisma) => {
```

Comment:
The magic is that it does not add any round-trip. It uses the underlying impl. of LRTs *but* since we don't call `await`, we are able to batch it. If you have a middleware, however, it will do that since you are requesting the result of the next one - can't get around that as I explained before.

Reply:
However, if you have a middleware with `await next...`, it will do that since you are requesting the result of the next one - can't get around that as I explained before. Our `DataLoader` batches all the queries that happen on the same `tick`... and `await` is causing the queries not to be on the same `tick`. So your while issue with rollbacks is solved, I'm sure that you would still sure that you'll like better batching.

Reply:
However, if you have a middleware with `await next...`, it will do that since you are requesting the result of the next one - can't get around that as I explained before. Our `DataLoader` batches all the queries that happen on the same `tick`... and `await` is causing the queries not to be on the same `tick`.

Reply:
I think that this is a good moment to try to change that since we're shipping a this behind a preview flag. I'm thinking of not batching per `tick`, but per time frame (eg. `20ms`) so that we can optimize your use-case. TBI

Reply:
Interesting on the batching, I was not aware of that.
My main concern was more that it would be another network call (read another potential for network delay/failure) that could affect currently working code (aka lets try not to disrupt good working production code).

As for batching, I think it probably could support both (basically switch between setTimeout and setImmediate) and offer the choice to the user.
This would be similar has to what GCP PubSub does things, you can control the time, amount of items, total size of payload, etc. but they offer sensible defaults. In general I prefer to offer knobs to advanced users (and I am pushing prisma to adopt that mentality since big co will want that anyway).

Reply:
A forced default for the middlewares + batching use-case would be that we automatically use `setTimeout` to save on round trips. On regular batches (no middlewares), this would just delay the queries, so not interesting there.